src/ripple/app/tx/impl/CancelTicket.cpp

```diff
@@ -41,49 +42,23 @@ class CancelTicket
 
     }
 
-    TER doApply () override
+    TER preCheck () override
     {
-        assert (mTxnAccount);
-
-        uint256 const ticketId = mTxn.getFieldH256 (sfTicketID);
-
-        SLE::pointer sleTicket = mEngine->view ().entryCache (ltTICKET, ticketId);
-
-        if (!sleTicket)
-            return tecNO_ENTRY;
-
-        auto const ticket_owner =
-            sleTicket->getFieldAccount160 (sfAccount);
-
-        bool authorized =
-            mTxnAccountID == ticket_owner;
-
-        // The target can also always remove a ticket
-        if (!authorized && sleTicket->isFieldPresent (sfTarget))
-            authorized = (mTxnAccountID == sleTicket->getFieldAccount160 (sfTarget));
+        // A Ticket may not be used to cancel a Ticket.  So a Sequence of
+        // zero (which indicates that a Ticket is being used) is malformed.
+        if (mTxn.getSequence () == 0)
+            return tefPAST_SEQ;
 
-        // And finally, anyone can remove an expired ticket
-        if (!authorized && sleTicket->isFieldPresent (sfExpiration))
-        {
-            std::uint32_t const expiration = sleTicket->getFieldU32 (sfExpiration);
-
-            if (mEngine->getLedger ()->getParentCloseTimeNC () >= expiration)
-                authorized = true;
-        }
-
-        if (!authorized)
-            return tecNO_PERMISSION;
-
-        std::uint64_t const hint (sleTicket->getFieldU64 (sfOwnerNode));
-
-        TER const result = mEngine->view ().dirDelete (false, hint,
-            getOwnerDirIndex (ticket_owner), ticketId, false, (hint == 0));
+        return Transactor::preCheck ();
+    }
 
-        adjustOwnerCount(mEngine->view(), mEngine->view().entryCache(
-            ltACCOUNT_ROOT, getAccountRootIndex(ticket_owner)), -1);
-        mEngine->view ().entryDelete (sleTicket);
+    TER doApply () override
+    {
+        assert (mTxnAccount);
 
-        return result;
+        // See if we can legitimately use the Ticket.  Note that canceling
+        // an expired Ticket is a tesSUCCESS.
+        return checkAndConsumeCancelTicket (mTxn, mTxnAccountID, mEngine);
```

Comment:
I struggled a bit with the three circumstances where we consume tickets: a) replacement for Sequence in a transaction, b) explicit cancel of the ticket, and c) consume the Ticket for a `tec` error.  They felt like they had a lot of commonality.  But the specifics of each case gave each a unique twist.

Anyway, I ended up making three different free functions that share a number of helper functions and a helper struct.  All three of the free function implementations are in the same file so they can share the helpers and so the helpers are readily accessible to anyone that maintains the code.

I think it is helpful to keep `checkAndConsumeSeqTicket()` and `checkAndConsumeCancelTicket()` in the same file and adjacent to one another.  It makes it easier for a maintainer to see that there are two different situations for removing tickets, and makes it easier to see the differences and commonalities.  As Nik suggests in a different place, I'll probably get rid of the `consumeTicketForTecError()` free function and inline that the one place it is called.

If you folks have suggestions for better names for the free functions I'm all ears.  I'm not at all wedded to the names.  Perhaps `checkAndConsumeTicket()` and `checkAndCancelTicket()` are a bit better?